Autonomous problem fix

ioshm opened the pull request.

Commit:
Upgrade to net9.0 and to the latest packages

MarkKharitonov commented on src/ZString.Unity/Assets/Scripts/ZString/FormatHelper.cs:
```diff
@@ -1,4 +1,4 @@
-﻿using System;
+using System;
```
This is noise in the delta. BOM must not be added or removed.

MarkKharitonov commented on src/ZString.Unity/Assets/Scripts/ZString/FormatHelper.cs:
```diff
@@ -48,17 +48,17 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
             {
                 var s = Unsafe.As<string>(arg);
-                int padding = width - s.Length;
+                int padding = width - s!.Length;
```
The change is correct, but should be applied at the point where `s` is initialized, like this - `var s = Unsafe.As<string>(arg)!;`. That way there would be no need to modify all the usages of `s`.

ioshm replied:
The compiler is smart enough to propagate the nullness property of the variable after the null forgiveness operator is applied to it. It is also smart enough to figure out that the null check should happen on the first use of the variable along an execution path.

```csharp
byte[]? s = null;
s[0] = 1; // Warning here.
s[1] = 1; // No warning here.
```

```csharp
byte[]? s = null;
s![0] = 1;
s[1] = 1;
```

But it seems that model is unaware of this fact, and liters the code with null forgiveness operator.

MarkKharitonov commented on src/ZString.Unity/Assets/Scripts/ZString/FormatHelper.cs:
```diff
@@ -48,17 +48,17 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
-                var span = sb.GetSpan(s.Length);
-                s.AsSpan().CopyTo(span);
-                sb.Advance(s.Length);
+                var span = sb.GetSpan(s!.Length);
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString.Unity/Assets/Scripts/ZString/FormatHelper.cs:
```diff
@@ -48,17 +48,17 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
-                s.AsSpan().CopyTo(span);
-                sb.Advance(s.Length);
+                var span = sb.GetSpan(s!.Length);
+                s!.AsSpan().CopyTo(span);
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString.Unity/Assets/Scripts/ZString/FormatHelper.cs:
```diff
@@ -48,17 +48,17 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
-                sb.Advance(s.Length);
+                var span = sb.GetSpan(s!.Length);
+                s!.AsSpan().CopyTo(span);
+                sb.Advance(s!.Length);
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString.Unity/Assets/Scripts/ZString/FormatHelper.cs:
```diff
@@ -131,15 +131,15 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
             {
                 var s = Unsafe.As<string>(arg);
-                int padding = width - s.Length;
+                int padding = width - s!.Length;
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString.Unity/Assets/Scripts/ZString/FormatHelper.cs:
```diff
@@ -131,15 +131,15 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
                 }
 
-                ZString.AppendChars(ref sb, s.AsSpan());
+                ZString.AppendChars(ref sb, s!.AsSpan());
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString.Unity/Assets/Scripts/ZString/FormatHelper.cs:
```diff
@@ -169,4 +169,4 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
 
-}
\ No newline at end of file
+}
```
Noise in the delta. Trailing newline must not be removed or added.

MarkKharitonov commented on src/ZString.Unity/Assets/Scripts/ZString/ZString.cs:
```diff
@@ -1,4 +1,4 @@
-﻿using System;
+using System;
```
This is noise in the delta. BOM must not be added or removed.

MarkKharitonov commented on src/ZString/FormatHelper.cs:
```diff
@@ -1,4 +1,4 @@
-﻿using System;
+using System;
```
This is noise in the delta. BOM must not be added or removed.

MarkKharitonov commented on src/ZString/FormatHelper.cs:
```diff
@@ -48,17 +48,17 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
             {
                 var s = Unsafe.As<string>(arg);
-                int padding = width - s.Length;
+                int padding = width - s!.Length;
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString/FormatHelper.cs:
```diff
@@ -48,17 +48,17 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
-                var span = sb.GetSpan(s.Length);
-                s.AsSpan().CopyTo(span);
-                sb.Advance(s.Length);
+                var span = sb.GetSpan(s!.Length);
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString/FormatHelper.cs:
```diff
@@ -48,17 +48,17 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
-                s.AsSpan().CopyTo(span);
-                sb.Advance(s.Length);
+                var span = sb.GetSpan(s!.Length);
+                s!.AsSpan().CopyTo(span);
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString/FormatHelper.cs:
```diff
@@ -48,17 +48,17 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
-                sb.Advance(s.Length);
+                var span = sb.GetSpan(s!.Length);
+                s!.AsSpan().CopyTo(span);
+                sb.Advance(s!.Length);
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString/FormatHelper.cs:
```diff
@@ -131,15 +131,15 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
             {
                 var s = Unsafe.As<string>(arg);
-                int padding = width - s.Length;
+                int padding = width - s!.Length;
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString/FormatHelper.cs:
```diff
@@ -131,15 +131,15 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
                 }
 
-                ZString.AppendChars(ref sb, s.AsSpan());
+                ZString.AppendChars(ref sb, s!.AsSpan());
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString/FormatHelper.cs:
```diff
@@ -169,4 +169,4 @@ private static void FormatToRightJustify<TBufferWriter, T>(ref TBufferWriter sb,
 
-}
\ No newline at end of file
+}
```
Noise in the delta. Trailing newline must not be removed or added.

MarkKharitonov commented on src/ZString/Utf16ValueStringBuilder.cs:
```diff
@@ -1,4 +1,4 @@
-﻿using System;
+using System;
```
This is noise in the delta. BOM must not be added or removed.

MarkKharitonov commented on src/ZString/Utf16ValueStringBuilder.cs:
```diff
@@ -119,7 +119,6 @@ public void Clear()
 
         public void TryGrow(int sizeHint)
         {
-
```
Noise in the delta. Unsolicited code formatting change - empty lines must be preserved.

MarkKharitonov commented on src/ZString/Utf16ValueStringBuilder.cs:
```diff
@@ -643,13 +642,13 @@ void AppendFormatInternal<T>(T arg, int width, ReadOnlySpan<char> format, string
                 {
                     var s = Unsafe.As<string>(arg);
-                    int padding = width - s.Length;
+                    int padding = width - s!.Length;
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString/Utf16ValueStringBuilder.cs:
```diff
@@ -643,13 +642,13 @@ void AppendFormatInternal<T>(T arg, int width, ReadOnlySpan<char> format, string
                     }
 
-                    Append(s);
+                    Append(s!);
```
Redundant when the initialization of `s` is adjusted - `var s = Unsafe.As<string>(arg)!;`

MarkKharitonov commented on src/ZString/Utf16ValueStringBuilder.cs:
```diff
@@ -717,7 +716,7 @@ static FormatterCache()
                 if (formatter == null)
                 {
-                    if (typeof(T).IsEnum)
+                    if (typeof(T).IsEnum && typeof(T).IsValueType)
```
An enum type is already a value type. This change is redundant.

MarkKharitonov commented on src/ZString/Utf16ValueStringBuilder.cs:
There are numerous changes in this file that change `buffer` to `buffer!`. All of them are correct, but redundant if `buffer` declaration is adjusted from `char[]? buffer;` to `char[] buffer;`

ioshm replied:
While true, the domain of the `buffer` field is indeed `byte[]?`. See the `Dispose` method which sets `buffer` to null and checks for it being `null`. I suppose ideally the methods would check if the struct was disposed, but that has performance implications.

Structs are also zero-init by default. For example `default(Utf16ValueStringBuilder).AppendLine();` will NRE because `buffer` is null.

I argue that it should be a nullable type since it captures the true domain of the variable, and that explicitly adding the null forgiveness operator at use sites captures the intent of ignoring this nullability. Of course deciding whether `buffer` should be a nullable type is at the discretion of the author/maintainer.

ioshm committed:
Fix nullability and generic constraints in EnumUtil<T>

- Added `where T : notnull` constraint to EnumUtil<T> to resolve CS8714 errors regarding nullability mismatch on generic type parameters.
- Used null-forgiveness operator (`!`) where needed to resolve possible null reference warnings (CS8602, CS8604).
- Ensured no behavior change, only nullability and constraint fixes.

ioshm committed:
Fix possible null reference dereference in FormatHelper.cs

- Added null-forgiveness operator (`!`) when using Unsafe.As<string>(arg) to resolve CS8602 warnings about possible null reference dereference.
- Ensured all string conversions and usages are null-forgiven where the logic expects non-null values.
- No behavior change, only null-safety improvements.

ioshm committed:
Fix nullability of parameters in ZStringWriter.cs

- Updated method signatures to match nullability of overridden members, e.g., `Write(string? value)` and `WriteAsync(string? value)`, to resolve CS8765 warnings.
- Used null-forgiveness operator where necessary.
- No behavior change, only null-safety and signature alignment.

ioshm committed:
Fix possible null reference dereference in Utf16ValueStringBuilder.cs

- Added null-forgiveness operator (`!`) to buffer and string usages where the logic guarantees non-nullness, resolving CS8602 warnings.
- No behavior change, only null-safety improvements.

ioshm committed:
Fix possible null reference dereference in Utf8ValueStringBuilder.cs

- Added null-forgiveness operator (`!`) to buffer and string usages where the logic guarantees non-nullness, resolving CS8602 and CS8604 warnings.
- No behavior change, only null-safety improvements.

ioshm committed:
Fix possible null reference return in ZString.cs

- Added null-forgiveness operator (`!`) to `Unsafe.As<string>(values[0])` in `JoinInternal<T>(ReadOnlySpan<char>, ReadOnlySpan<T>)` to resolve CS8603 warning about possible null reference return.
- No behavior change, only null-safety improvement.
